pkg/service-manager/clients/k8s_service_catalog.go

```diff
@@ -0,0 +1,436 @@
+	}
+	var serviceClasses []entitystore.Entity
+
+	for _, csc := range cscs {
+		log.Debugf("Syncing service class %s", csc.Spec.ExternalName)
+		if csc.Status.RemovedFromBrokerCatalog {
+			continue
+		}
+		plans, err := c.sdk.RetrievePlansByClass(&csc)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Error retreiving plan for service class %s", csc.Spec.ExternalName)
+		}
+		var serviceClassPlans []entities.ServicePlan
+		for _, plan := range plans {
+			if plan.Status.RemovedFromBrokerCatalog {
+				continue
+			}
+			marshall := func(ps *runtime.RawExtension) (*spec.Schema, error) {
+				out := new(spec.Schema)
+				if ps != nil && ps.Size() > 0 {
+					b, _ := json.Marshal(ps)
+					if err := json.Unmarshal(b, out); err != nil {
+						return nil, errors.Wrap(err, "could not decode schema")
+					}
+				}
+				return out, nil
+			}
+			create, err := marshall(plan.Spec.ServiceInstanceCreateParameterSchema)
+			if err != nil {
+				return nil, errors.Wrap(err, "Error marshalling create schema")
+			}
+			update, err := marshall(plan.Spec.ServiceInstanceUpdateParameterSchema)
+			if err != nil {
+				return nil, errors.Wrap(err, "Error marshalling update schema")
+			}
+			bind, err := marshall(plan.Spec.ServiceBindingCreateParameterSchema)
+			if err != nil {
+				return nil, errors.Wrap(err, "Error marshalling bind schema")
+			}
+
+			bindable := csc.Spec.Bindable
+			if plan.Spec.Bindable != nil {
+				bindable = *plan.Spec.Bindable
+			}
+
+			serviceClassPlans = append(serviceClassPlans, entities.ServicePlan{
+				BaseEntity: entitystore.BaseEntity{
+					Name:   plan.Spec.ExternalName,
+					Status: entitystore.StatusREADY,
+				},
+				PlanID:      plan.Name,
+				Description: plan.Spec.Description,
+				Schema: entities.Schema{
+					Create: create,
+					Update: update,
+					Bind:   bind,
+				},
+				Free:     plan.Spec.Free,
+				Bindable: bindable,
+				Metadata: plan.Spec.ExternalMetadata,
+			})
+		}
+		serviceClasses = append(serviceClasses, &entities.ServiceClass{
+			BaseEntity: entitystore.BaseEntity{
+				Name:   csc.Spec.ExternalName,
+				Status: entitystore.StatusREADY,
+			},
+			ServiceID: csc.Name,
+			Broker:    csc.Spec.ClusterServiceBrokerName,
+			Bindable:  csc.Spec.Bindable,
+			Plans:     serviceClassPlans,
+		})
+	}
+
+	return serviceClasses, nil
+}
+
+// ListServiceInstaces returns a list of ServiceInstance entities which correspond to the provisioned services.  Most
+// importantly the status of each service instance is returned as well.
+func (c *k8sServiceCatalogClient) ListServiceInstances() ([]entitystore.Entity, error) {
+	instances, err := c.sdk.RetrieveInstances(c.config.CatalogNamespace)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error retreiving service instance list")
+	}
+	var serviceInstances []entitystore.Entity
+	for _, instance := range instances.Items {
+
+		parameters := make(map[string]interface{})
+		err := json.Unmarshal(instance.Spec.Parameters.Raw, &parameters)
```

Comment:
Similar to `ServiceBinding` looks like the `parameters` are optional. So may be check if it's defined :-
https://github.com/kubernetes-incubator/service-catalog/blob/4a138279dbf7bbb3b3d0e51dcca50858e6738604/pkg/apis/servicecatalog/types.go#L599

Reply:
Yeah, I need to verify some of these conditions through unit tests too.  Another good catch.